Fix texture memory usage, by unloading tiles, resolving #12906

tristan-morris opened the pull request:
This PR resolves #12906, and most likely quite a few other issues relating to loss of WebGL context, along with excessive memory use and crashing.

The root cause is textures aren't being unloaded from **GL memory** when they're no longer needed. This is most pronounced when moving around the map, with `flyTo` being the most pathological case. From what I can tell, this impacts all platforms but particularly resource constrained devices. I'd guess Safari has lower memory limits than other browsers. 

For me, this showed up in Safari as the tab hanging, Safari killing the tab, iOS/macOS killing Safari, and occasionally iOS panicking. In my app, this showed as WkWebView being terminated. If you follow the device logs (not the app logs in XCode) it's possible to see warnings related to excessive memory usage. But, the excessive memory use isn't specifically attributed to VRAM usage! I might raise this on their Bugzilla.

Before:
Top left shows an overlay with predicted vram usage in MB. By 40 seconds, Safari has terminated the tab, so ~7 pans/zooms.

https://github.com/mapbox/mapbox-gl-js/assets/44310937/c42891c8-ea4f-42ea-97cf-03408a4921f1


After:
Max GL memory reaches ~210MB, with satellite raster and DEM. Safai seems happy. 

https://github.com/mapbox/mapbox-gl-js/assets/44310937/e695d573-e087-4786-94ad-a5c589806fe9

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->

 - [X ] briefly describe the changes in this PR
 - [ X] include before/after visuals or gifs if this PR includes visual changes
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ? kind of] post benchmark scores
 - [X ] manually test the debug page
 - [ ] tagged `@mapbox/map-design-team` `@mapbox/static-apis` if this PR includes style spec API or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [ ] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [X ] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Fix an issue with excessive video memory usage with raster sources, as textures weren't being unloaded from video memory when no longer needed. </changelog>`
@

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mapbox/mapbox-gl-js?pullRequest=12924) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/mapbox/mapbox-gl-js?pullRequest=12924) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mapbox/mapbox-gl-js?pullRequest=12924) it.</sub>

mourner commented:
@tristan-morris this is really impressive work, thanks so much for taking on this! Our memory handling definitely needs this kind of tightening up, and it's been a while since we did such a thorough hunt for memory leaks.

For a little bit of background on texture retention — originally it was introduced 9 years ago (!) in #597 for performance reasons described in #591 — to have a pool of common size textures to reuse instead of continuously creating and deleting textures. Before we remove this mechanism, let's investigate to understand why exactly it leads to unbounded memory growth — in theory, if I remember correctly, the number of saved textures should be bounded by what's seen on screen, and any newly visible tiles should reuse textures left in the pool for those that went offscreen, thus balancing the total amount. Did we break that balance at some point? Or is the amount retained too much even if it's bounded, and it's better for performance to always delete and create new textures without reuse?

tristan-morris commented:
Thanks @mourner. More than happy to help and it's been a good opportunity to understand the codebase in a lot more detail. The background is really useful and makes a lot of sense. 

My takeaway is the texture handling has evolved over time and for the different concerns you've outlined, this has somewhat dropped off, and become quite fragmented. This can be seen in `tile.js` in particular with all of the different variables utilising `Texture`--all super important I bet, but the particular texture caching path I removed was only used by the raster layer source. 

My thinking so far is to find the last bits of code which are holding onto textures quite aggressively, and then re-optimise texture allocations with reuse across everything utilising a texture? There might be a bit of a resource trade-off here between CPU/GPU and VRAM allocation. 

As a comparison, if I load the pathological-flyto.html sample which utilises the MB Earthquake example, the texture usage is ~72MB and then in my app it's ~170MB. Similar enough, but I'm loading sig more data across ~7-8 sources. If I zoom in and out a few times on the same location with geojson loaded, this balloons to 1GB and 223MB, respectively. I think this high point is indicative of the texture caches being fully hydrated (tiles being retained due to current viewport). So overall, maybe slightly too much is being retained in VRAM but I guess this limit could be dynamic (ideally) and performance characteristics calculated per device.

tristan-morris commented:
I think this is ready for another look @mourner. 

It looks like different gl unload codepaths weren't releasing their resources consistently, which meant the texture/raster tile cache could continually grow.  There was an unloadTileData and an unloadTile call which would have pushed twice into the cache, the same texture I guess but then it would have been popped once. Then again (eventually) perhaps causing some weird texture changes(?) but I wonder if this quirk is worked around elsewhere so it's not obvious, or an unload check. Either way, it causes the cache to grow. 

The solution I implemented was to add types, and then destroy everything with the same Tile.destroy() method. The tile cache has remained pretty stable since (e.g. instrumenting the cache to check growth). I did have a fixed cap, but I think if it's all correctly pushed/popped, as you said, it should be fine. Feels slightly diiirty without a hard cap but ok. I refactored other cleanup logic so things aren't as easily missed. 


This becomes most problematic on Webkit, iOS, with dem or raster and doing more with the library (consuming higher mems). Higher dpi devices (iPhone etc.) also fetching the 2x textures chew more allocation too. So you kind of end up with the absolute worst case user - an iPhone target, with an allocator that behaves like a potato. 

With the changes:
Main ~50MB stable.
Workers 16MB stable.
GPU 300MB with Satellite, terrain etc. stable.

I'll write up a separate issue with larger geojson data sources running on a potato.

mourner commented:
@tristan-morris this is fantastic work, thanks so much for working further on this! I'm pretty sure we'll try to land this before v3 final — let me spend some time reviewing some details, but overall this looks great.

mourner commented on src/source/raster_tile_source.js:
```diff
@@ -221,10 +222,9 @@ class RasterTileSource extends Evented implements Source {
-        if (tile.texture) {
-            painter.saveTileTexture(tile.texture);
-        }
+        // Texture caching on unload occurs in unloadTile
```
@stepankuzmin can you please check the logic here re `CustomSource`? Looks like the way this is currently wired, after the change, it would stop caching textures — what's the best way to accommodate the new approach while making sure nothing regresses there?

stepankuzmin replied:
Hey all! I've checked the CustomSource caching, and it was affected by this PR. I've made a fix for that in https://github.com/mapbox/mapbox-gl-js/pull/12993. Since this PR was made from a fork, I couldn't target my changes on top of it. We can merge them sequentially. /cc @mourner

mourner commented on src/source/raster_tile_source.js:
```diff
@@ -238,7 +238,20 @@ class RasterTileSource extends Evented implements Source {
+            tile.destroy(true);
+
+            // Save the texture to the cache
+            if (tile.texture && tile.texture instanceof Texture) {
```
Nit: redundant check

tristan-morris replied:
I thought that too! Seems that Flow picks up that `tile` may have been modified

mourner replied:
In that case you can usually placate Flow by separating the condition (or a texture reference) into a variable.

mourner commented on src/source/raster_tile_source.js:
```diff
@@ -238,7 +238,20 @@ class RasterTileSource extends Evented implements Source {
+        } else {
+            tile.destroy();
+        }
+
```
Just wondering — would it simplify things if we moved this whole logic to `tile.destroy` instead of having a conditional parameter? It might be better for consistency, since the counterpart `getTileTexture` logic of reusing textures is inside `tile.js` (`setTexture` method).

tristan-morris replied:
I was thinking that too. My thinking against was, if you call destroy() on something you think it should be destroyed. Or at least provide a mechanism which makes sense. I also didn't want to move the calls around too much :)

mourner commented on src/terrain/draw_terrain_raster.js:
```diff
@@ -331,7 +335,11 @@ function drawTerrainRaster(painter: Painter, terrain: Terrain, sourceCache: Sour
                 context.activeTexture.set(gl.TEXTURE0);
-                tile.texture.bind(gl.LINEAR, gl.CLAMP_TO_EDGE, gl.LINEAR_MIPMAP_NEAREST);
+                if (tile.texture) {
+                    // This call seemed invalid with 3 args?
+                    // tile.texture.bind(gl.LINEAR, gl.CLAMP_TO_EDGE, gl.LINEAR_MIPMAP_NEAREST);
```
Seems to be accidental, let's just drop this.
```suggestion
```

mourner commented on src/render/program/line_program.js:
```diff
@@ -93,7 +93,7 @@ const lineUniformValues = (
         'u_gradient_image': 1,
         'u_image_height': imageHeight,
-        'u_texsize': hasDash(layer) ? tile.lineAtlasTexture.size : [0, 0],
+        'u_texsize': hasDash(layer) ? (tile.lineAtlasTexture ? tile.lineAtlasTexture.size : [0, 0]) : [0, 0],
```
Nit
```suggestion
        'u_texsize': hasDash(layer) && tile.lineAtlasTexture ? tile.lineAtlasTexture.size : [0, 0],
```

mourner reviewed (approved):
👍 given #12993 will be merged afterwards to address `CustomSource` caching.
